fix(40454): allow IntersectionObserverInit.root to accept type Docume…

watrool opened the pull request:
…nt and update IntersectionObserver.root to be of type Node

Fix https://github.com/microsoft/TypeScript/issues/40454

watrool commented on baselines/dom.generated.d.ts:
```diff
@@ -9607,7 +9607,7 @@ declare var InputEvent: {
 /** provides a way to asynchronously observe changes in the intersection of a target element with an ancestor element or with a top-level document's viewport. */
 interface IntersectionObserver {
-    readonly root: Element | null;
+    readonly root: Node | null;
```
![image](https://user-images.githubusercontent.com/7275468/92752827-4a271100-f357-11ea-9a0c-0328981a0324.png)

watrool commented on baselines/dom.generated.d.ts:
```diff
@@ -611,7 +611,7 @@ interface IntersectionObserverEntryInit {
 
 interface IntersectionObserverInit {
-    root?: Element | null;
+    root?: Document | Element | null;
```
![image](https://user-images.githubusercontent.com/7275468/92752950-64f98580-f357-11ea-8df9-fbec78a06297.png)

saschanaz commented:
The right way to do this is updating `idlSource.json` to point https://w3c.github.io/IntersectionObserver/ and running `npm run fetch-idl "Intersection Observer" && npm run build && npm run baseline-accept`, so that we can keep syncing with the upstream changes.

watrool commented:
@saschanaz thanks for the comment! I tried that approach, but it seems `https://w3c.github.io/IntersectionObserver/ ` changed all `void` type to `undefined` type 
![image](https://user-images.githubusercontent.com/7275468/92793795-b700d200-f37c-11ea-8b74-3a13436de712.png)
which makes the script throws
```
Error: Unknown DOM type: undefined
    at convertDomTypeToTsTypeSimple (/Users/troytao/TSJS-lib-generator/lib/emitter.js:301:15)
    at convertDomTypeToTsTypeWorker (/Users/troytao/TSJS-lib-generator/lib/emitter.js:239:28)
    at convertDomTypeToTsType (/Users/troytao/TSJS-lib-generator/lib/emitter.js:233:22)
    at emitSignature (/Users/troytao/TSJS-lib-generator/lib/emitter.js:598:26)
    at /Users/troytao/TSJS-lib-generator/lib/emitter.js:610:45
    at Array.forEach (<anonymous>)
    at emitSignatures (/Users/troytao/TSJS-lib-generator/lib/emitter.js:610:30)
    at emitMethod (/Users/troytao/TSJS-lib-generator/lib/emitter.js:590:13)
    at /Users/troytao/TSJS-lib-generator/lib/emitter.js:620:31
    at Array.forEach (<anonymous>)
```

Is it ok if I add a `['undefined', 'void']` to `baseTypeConversionMap` in `emitter.ts` to fix this ?

saschanaz commented:
> Is it ok if I add a `['undefined', 'void']` to `baseTypeConversionMap` in `emitter.ts` to fix this ?

Gread find! Yes, that will fix it.

watrool commented:
@saschanaz updated!

saschanaz reviewed (approved):
LGTM 👍

watrool commented:
@sandersn poke you for merging this PR ...

saschanaz commented on src/helpers.ts:
```diff
@@ -16,7 +16,8 @@ export const baseTypeConversionMap = new Map<string, string>([
     ["FrozenArray", "ReadonlyArray"],
-    ["EventHandler", "EventHandler"]
+    ["EventHandler", "EventHandler"],
+    ["undefined", "void"]
```
BTW, #920 did this in another way so we don't need this anymore. Sorry, but still was a good work!

saschanaz commented:
Could you run `npm run fetch-idl "Intersection Observer" && npm run build && npm run baseline-accept` once again? I think that will fix the attribute type also to `Element | Document | null`.

watrool commented:
@saschanaz thanks for notifying! I just rebased the PR, removed the change in `helpers.ts` and ran `npm run fetch-idl "Intersection Observer" && npm run build && npm run baseline-accept`.

saschanaz commented:
The TypeScript team will review this when the "review time" comes, so please wait until that happens 🙏
